### PR TITLE
daemon: release sandbox even when NetworkDisabled

### LINKED ISSUE #42461: Bug: docker hang on restarting when bridge set to none.
**Description**
Starting container will create sandbox still, although we set bridge none on config.
The sandbox won't be deleted until restarting the docker.
If create a lot containers, dockerd will hang on restarting.

**Steps to reproduce the issue:**
1. Update the docker config set "bridge": "none"
2. Use for i in $(seq 1 4000); do docker run --rm ubuntu; done to create many sandbox.
3. Restart docker and wait.(We can verify it by read network local-kv.db directly.)

**Describe the results you received:**
Dockerd will hang on restarting. We can know the restore goroutine stuck on [CleanSandbox()](https://github.com/moby/moby/blob/7c6a9484ee37b28fa314f716581bdb268cd78ce2/vendor/github.com/docker/libnetwork/controller.go#L248) from stackinfo.

**Describe the results you expected:**
Don't create sandbox if needless.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
docker 19.03.9
```

### LINKED ISSUE #46583: Stale network sandboxes causing long docker startup time
### Description

When you have the docker daemon configured with "--bridge none", and you launch containers via "docker run" without passing a --network option, then stale network sandbox accumulates, and on the next restart of the docker daemon, docker will remove these stale network sandboxes and this can substantially slow docker startup time when you have a lot of these or when docker is running with limited resources.

I started investigating this issue since at some point I had a machine where docker took quite a minutes to start and I was wondering what was going on.

Granted, it might be considered a bit bogus to have the combination of "--bridge none" on the docker daemon and run (some) containers without specifying the --network option, which is something I've adjusted (to explicitly pass `--network none`). But the docker behavior seems a bit misleading and I wonder if this doesn't hide another issue, where network sandboxes should not become stale.

### Reproduce

Given I have the following in /etc/docker/daemon.json
```
{
    "bridge": "none"
}
```
And I restart docker
And I launch a few containers using something like:
```
for i in $(seq 50); do docker run --rm busybox echo $i; done
```
When I restart docker
Then I see in the docker logs 50 occurrences of `Removing stale sandbox <sandbox ID> (<container ID>)` (e.g. ""Removing stale sandbox d7548f4571ba16a4b925a11b75088f36148164993e43085c680075257dbb1c9f (9a64aa6aeadc4
817088db55c6ac520b69e275e248045d895ee6fac283e762823)"
And ultimately docker takes more time to start (this is particularly visible on low-power machine and when there's many (>1000) stable sandboxes)

### Expected behavior

No stale sandbox removed when docker is restarted.

### docker version

```bash
Client: Docker Engine - Community
 Version:           24.0.6
 API version:       1.43
 Go version:        go1.20.7
 Git commit:        ed223bc
 Built:             Mon Sep  4 12:32:10 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          24.0.6
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.7
  Git commit:       1a79695
  Built:            Mon Sep  4 12:32:10 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.24
  GitCommit:        61f9fd88f79f081d64d6fa3bb1a0dc71ec870523
 runc:
  Version:          1.1.9
  GitCommit:        v1.1.9-0-gccaecfc
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client: Docker Engine - Community
 Version:    24.0.6
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.11.2
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.21.0
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 1
  Running: 0
  Paused: 0
  Stopped: 1
 Images: 5
 Server Version: 24.0.6
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 61f9fd88f79f081d64d6fa3bb1a0dc71ec870523
 runc version: v1.1.9-0-gccaecfc
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.1.0-12-amd64
 Operating System: Debian GNU/Linux 12 (bookworm)
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 724.1MiB
 Name: debian
 ID: 175bab96-368d-4b39-8d68-57ed259ab284
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by akerouanton**
- Fix https://github.com/moby/moby/issues/42461
- Fix https://github.com/moby/moby/issues/46583
- Closes https://github.com/moby/moby/pull/42454. Commit message has been improved and a couple comments have been added.

**- What I did**

When the default bridge is disabled by setting dockerd's `--bridge=none` option, the daemon still creates a sandbox for containers with no network attachment specified. In that case `NetworkDisabled` will be set to true.

However, currently the `releaseNetwork` call will early return if NetworkDisabled is true. Thus, these sandboxes won't be deleted until the daemon is restarted. If a high number of such containers are created, the daemon would then take few minutes to start.

As a side note, `NetworkDisabled` semantics is weird/broken and should be revised:

* On one hand a sandbox is created even if `NetworkDisbled` is set. Thus it allows these containers to be manually connected to other networks;
* OTOH, when manually connecting such container to a network nothing happens and no error is returned (ie. no interface and no route provisioned, no embedded DNS, etc...);

**- Description for the changelog**

* Fix a bug that would prevent network sandboxes to be properly deleted when stopping containers with no network attachment are specified and dockerd's `--bridge=none` option is specified.

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
> OTOH, when manually connecting such container to a network nothing happens and no error is returned (ie. no interface and no route provisioned, no embedded DNS, etc...);

Yeah, that's confusing. 

I also recall having discussions about `--network=none` and if it would be _technically_ possible to connect a container to a network "after the fact". I.e., we currently don't allow a container to be connected to a network once it's created with `none` network;


```bash
docker network create foo

# this creates a container with a sandbox, but no networking connection
docker run -d --network=none --name hello nginx:alpine

# and connecting to a network produces an error
docker network connect foo hello
Error response from daemon: container cannot be connected to multiple networks with one of the networks in private (none) mode
```

I could see use-cases where you'd want to create a container without networking (at first), and later connect it to a network, but not sure what'd be involved in doing so (if possible!)

File: daemon/container_operations.go [hunk 1]
```diff
@@ -967,10 +967,17 @@ func (daemon *Daemon) getNetworkedContainer(containerID, connectedContainerID st
 
 func (daemon *Daemon) releaseNetwork(container *container.Container) {
 	start := time.Now()
+	// If live-restore is enabled, the daemon cleans up dead containers when it starts up. In that case, the
+	// netController hasn't been initialized yet and so we can't proceed.
+	// TODO(aker): If we hit this case, the endpoint state won't be cleaned up (ie. no call to cleanOperationalData).
 	if daemon.netController == nil {
```
Comment (thaJeztah) [inline daemon/container_operations.go:973]: Wondering if there's an actual case where `daemon.netController` _CAN_ be `nil` 🤔

I'm considering that that should (probably?) never happen, and if it _would_ we should probably make that a fatal error, and exit (before we even reach this code)

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
should we cherry-pick this fix?

File: daemon/container_operations.go (hunk 1 above)
Comment (akerouanton) [reply daemon/container_operations.go:973]: Oh actually my comment isn't about an hypothetical situation:

* `(*Daemon).releaseNetwork` is called by `(*Daemon).Cleanup`: https://github.com/moby/moby/blob/670bc0a46c4ca03b75f1e72f73b8302bce90275a/daemon/start.go#L243
* Which is called by `(*Daemon).restore` on L449: https://github.com/moby/moby/blob/670bc0a46c4ca03b75f1e72f73b8302bce90275a/daemon/daemon.go#L449
* But `netController` is initialized by calling `(*Daemon).initNetworkingController` on L530: https://github.com/moby/moby/blob/670bc0a46c4ca03b75f1e72f73b8302bce90275a/daemon/daemon.go#L530

**Comment by thaJeztah** [comment]
@neersighted LGTY?

**Review by corhere — APPROVED** [review]

**Comment by thaJeztah** [comment]
Let me bring this one in; just wanted to have a second set of eyes before merging 👍

**Pull request merged**